### PR TITLE
Add the TestEnableDevelopmentNetworkProtocols configuration flag

### PR DESCRIPTION
**Pull request by nfrisby**
Fixes  [CAD-2690.](https://jira.iohk.io/browse/CAD-2696)

### CODE REVIEW

**Comment by nfrisby** [comment]
The latest in this list of arguments (currently Mary) may also need to depend on this new flag, at different points in the future. Right now, it doesn't need to.

https://github.com/input-output-hk/cardano-node/blob/58c1e06f3618b546627916bd24649d5cecf63ff8/cardano-node/src/Cardano/Node/Protocol/Cardano.hs#L188-L195

File: cabal.project
Comment (nfrisby) [inline cabal.project:163]: This will have to change again once we merge the corresponding PR in `ouroboros-network` repo.

**Comment by nfrisby** [comment]
This currently depends on the upstream PR input-output-hk/ouroboros-network#2947

File: cardano-node/src/Cardano/Node/Configuration/POM.hs
Comment (coot) [inline cardano-node/src/Cardano/Node/Configuration/POM.hs:67]: `ncEnableDevelopmentVersions` sounds better for me, but I am not native speaker.  Should I adjust my ear? Or both versions are acceptable?

File: cabal.project
Comment (coot) [reply cabal.project:163]: Is this the https://github.com/input-output-hk/ouroboros-network/pull/2947 PR in question?

**Review by coot — APPROVED** [review]
This PR looks good to me, I also left comments on input-output-hk/ouroboros-network#2497.

File: cardano-node/src/Cardano/Node/Configuration/POM.hs
Comment (nfrisby) [reply cardano-node/src/Cardano/Node/Configuration/POM.hs:67]: @coot I was having a hard time choosing between those two. I think both are OK, neither is great.

I chose "developmental version" because "development version" feels slightly more ambiguous: I think it's a bit more likely that someone would unfortunately interpret "development version" as "the version I should use for development" or something like that. Even though -- in a vague sense -- that's not wrong.

But: I don't think either name is great. Do you have any fresh ideas for something that's clear?

  * experimental version
  * testing version
  * quality assurance version
  * in-development version
  * etc?

Comment (coot) [reply cardano-node/src/Cardano/Node/Configuration/POM.hs:67]: I like both `in-development` and `experimental` version, another one could be `future-version`.  Maybe the most explicit is `experimental`, what do you think?

Comment (nfrisby) [reply cardano-node/src/Cardano/Node/Configuration/POM.hs:67]: I just pinged Duncan and he expressed a preference for "Test*", since that's how this flag's ancestors have been named.

So I'm thinking `TestInDevelopmentVersion`.

(PS - Woohoo Mary hard fork!)

**Comment by nfrisby** [comment]
I think input-output-hk/ouroboros-network#2947 is now ready for merge; it's awaiting Approval.

Once that happens, I can bump that entry in `cabal.project` and then I think this PR is also ready for merge.

cc: @coot @Jimbo4350

**Comment by nfrisby** [comment]
> I think [input-output-hk/ouroboros-network#2947](https://github.com/input-output-hk/ouroboros-network/pull/2947) is now ready for merge; it's awaiting Approval.
> 
> Once that happens, I can bump that entry in `cabal.project` and then I think this PR is also ready for merge.

We just merged it, and I've just now pushed the update. I don't know if this interferes with the other `cabal.project` updating that eg @newhoggy are doing -- @Jimbo4350 do you have clarity there?

Clarification: that upstream commit does not yet filter out any versions, so this option is currently a noop with respect to the node behavior.

I'm changing this PR from Draft PR to a ready-to-merge-pending-Approval PR.

**Review by newhoggy — APPROVED** [review]

**Comment by nfrisby** [comment]
@Jimbo4350 @newhoggy Please merge this whenever you see fit, since there are similar `cabal.project` PRs in flight. Thanks!

**Review by dcoutts — APPROVED** [review]
To be rebased and merged after https://github.com/input-output-hk/cardano-node/pull/2657 is merged.

**Comment by dcoutts** [comment]
Note that only the last 4 patches are really part of this PR, the earlier ones are from https://github.com/input-output-hk/cardano-node/pull/2657

**Comment by dcoutts** [comment]
BTW, I took the liberty of renaming the option to `TestEnableDevelopmentNetworkProtocols` and then adding the similarly-named one `TestEnableDevelopmentHardForkEras` to the same PR. The latter is as the name suggests, for advertising support for new eras.

File: cardano-node/src/Cardano/Node/Configuration/POM.hs
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Node/Configuration/POM.hs]: :+1:

File: cardano-node/src/Cardano/Node/Configuration/POM.hs
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Node/Configuration/POM.hs:170]: :+1:

**Review by Jimbo4350 — APPROVED** [review]
LGTM!

**Pull request merged**
